### PR TITLE
Method returns a single argument

### PR DESCRIPTION
**Pull request by zdevito**
This PR changes Method (just Method not all graphs) to always have a single
return argument. 

This is part 1 in a set of changes that will enable us to have better handling if early return statements.
The simplification that this change provides greatly reduces the work for the next step.

This change makes it so that Method and Python handle multiple returns in the same way:
* 0 - None
* 1 - <single value>
* many - Tuple[...]

The result is that a lot of special-case handling in compiler.cpp and its
bindings can be removed. It also fixes several bugs in return handling,
including one where return values were not always checked against their
attributed values.

Notes:
* inferTypeFrom is renamed to be more accurate and discourage use.
* This has uncovered some bugs in other components, which are noted in
  the diff.

### CODE REVIEW

File: torch/csrc/jit/script/compiler.cpp [hunk 1]
```diff
@@ -1035,50 +1041,27 @@ struct to_ir {
...
-      }
-    } else if (schema.returns().size() > 0) {
-      // schema has returns but there's no return nodes in graph
-      throw ErrorReport() << "Expected " << schema.returns().size()
-                          << " return value"
-                          << (schema.returns().size() > 1 ? "s" : "")
-                          << " but found no return statement";
+    Value* result = return_stmt ? emitExpr(return_stmt->expr())
+                                : graph->insertConstant(IValue(), range);
+    TypePtr result_type = schema.returns().size() > 0
```
Comment (suo) [inline torch/csrc/jit/script/compiler.cpp]: Confused by this logic. If a schema has multiple returns, why is `result_type` set to the type of the first one?

File: torch/csrc/jit/script/compiler.cpp [hunk 2]
```diff
@@ -1035,50 +1041,27 @@ struct to_ir {
...
+          range, *graph, result_type, result, /*allow_conversions=*/true);
+    }
+
+    if (!result->type()->isSubtypeOf(result_type)) {
+      throw ErrorReport(range) << "Return value was annotated as having type " << result_type->python_str()
+        << " but is actually of type " << result->type()->python_str();
     }
-    return returns;
+    graph->registerOutput(result);
+    return Argument("", result_type);
```
Comment (wanchaol) [inline torch/csrc/jit/script/compiler.cpp]: It is weird that we treat return as a Argument

**Review by wanchaol — APPROVED** [review]
looks good

File: torch/csrc/jit/script/compiler.cpp (hunk 1 above)
Comment (wanchaol) [reply torch/csrc/jit/script/compiler.cpp]: I think schema can only have one return or None now as stated by the `JIT_ASSERT` stmt.

Comment (zdevito) [reply torch/csrc/jit/script/compiler.cpp]:
```
    // we represent no annoation on a return type as having no values in the
     // schema's return() list
     // in emitReturn we take the actual return value to be the value of the return
     // statement if no one was provided here
```
Ill add a back reference.

Comment (zdevito) [reply torch/csrc/jit/script/compiler.cpp]: In other words, we need an encoding for "user has not annotated the return type, infer it", and using schema.returns() being empty (versus having exactly 1 element) seemed to be the simplest thing. Previously this was encoded using `is_varret` but the implementation was buggy and it allowed incorrect types to be returned.

**Pull request closed without merging**